src/irmin-pack/irmin_pack.ml

```diff
@@ -557,8 +577,76 @@ struct
         in
         Contents.CA.sync ~on_generation_change (contents_t t)
 
-      (** Storea share instances so one clear is enough. *)
+      (** Stores share instances so one clear is enough. *)
       let clear t = Contents.CA.clear (contents_t t)
+
+      (** Migrate data from the IO [src] (with [name] in path [root_old]) into
+          the temporary dir [root_tmp], then swap in the replaced version. *)
+      let migrate_io_to_v2 ~root_old ~root_tmp ~progress (name, src) =
+        let old, tmp = (root_old // name, root_tmp // name) in
+        IO.upgrade ~progress ~src ~dst:(tmp, `V2) |> function
+        | Ok () ->
+            Platform.rename tmp old;
+            IO.close src;
+            ()
+        | Error (`Msg s) -> invalid_arg s
+
+      let migrate config =
+        let* () =
+          if readonly config then Lwt.fail RO_Not_Allowed else Lwt.return_unit
+        in
+        let config = Irmin.Private.Conf.add config readonly_key true in
+        Log.info (fun l -> l "[%s] migrate" (root config));
+        let* t = unsafe_v config in
+        match Contents.CA.version (contents_t t) with
+        | `V2 as v ->
+            Log.app (fun l ->
+                l "Store at %s is already in current version (%a)" (root config)
+                  pp_version v);
+            close t
+        | `V1 ->
+            let* () = close t in
+            let root_old = root config in
+            let root_tmp =
+              let rand =
+                Random.State.(bits (make_self_init ())) land 0xFFFFFF
+              in
+              root_old // Fmt.str "tmp-migrate-%06x" rand
+            in
+            Log.info (fun l ->
+                l "Creating temporary directory `%s' for the migrated store"
+                  root_tmp);
+            Platform.mkdir root_tmp;
+            let ios : (string * IO.t) list =
+              [ "store.pack"; "store.branches"; "store.dict" ]
+              |> List.map (fun name ->
+                     ( name,
+                       IO.v ~version:`V1 ~fresh:false ~readonly:true
+                         (root_old // name) ))
+            in
+            let total =
+              ios |> List.map (snd >> IO.offset) |> List.fold_left Int64.add 0L
+            in
+            let bar, progress =
+              Utils.Progress.counter ~total ~sampling_interval:10
+                ~message:"Migrating store" ~pp_count:Utils.pp_bytes ()
+            in
+            ios |> List.iter (migrate_io_to_v2 ~root_old ~root_tmp ~progress);
+            Utils.Progress.finalise bar;
+            Platform.rmdir root_tmp;
```

Comment:
The more POSIX calls we expose/use here the harder it will be to make a proper MirageOS port :p So same here, if we could get rid of that `rmdir` that would be great.

Reply:
Scrapped too: `tmp` files are now of the form `<srcname>-tmp-migrate-a1b2c31`, rather than using a sub-directory.